spark-cassandra-connector/src/it/scala/com/datastax/spark/connector/sql/CassandraSQLPredicatePushdownSpec.scala

```diff
@@ -0,0 +1,131 @@
+  }
+
+  it should "not allow predicate pushdown if previous cluster column has None-EQ predicate." +
+    " a=1 and b=1 and c=1 and d=3 and e=1 and f>2 are pushed down" in {
+    val result = cc.sql("select * from test_pd where a=1 and b=1 and c=1 and d =3 and e=1 and f>2 and g=3").collect()
+  }
+
+  it should "not allow predicate pushdown if previous cluster column has None-EQ predicate." +
+    " a=1 and b=1 and c=1 and d=3 and e>1 are pushed down" in {
+    val result = cc.sql("select * from test_pd where a=1 and b=1 and c=1 and d =3 and e>1 and f>2 and g=3").collect()
+  }
+
+  it should "allow predicate pushdown if partition column predicates can be pushed down and only last cluster column has IN predicate." +
+    " a=1 and b=1 and c=1 and d=3 and e=1 and f=1 and g=2 and h in (1, 2, 3) are pushed down" in {
+    val result = cc.sql("select * from test_pd where a=1 and b=1 and c=1 and d=3 and e=1 and f=1 and g=2 and h in (1, 2, 3)").collect()
+  }
+
+  it should "not allow predicate pushdown if previous cluster column has None-EQ predicate." +
+    " a=1 and b=1 and c=1 and d=3 and e=1 and  f>1 are pushed down" in {
+    val result = cc.sql("select * from test_pd where a=1 and b=1 and c=1 and d=3 and e=1 and f>1 and h in (1, 2, 3)").collect()
+  }
+
+  it should "allow predicate pushdown if there is EQ indexed column predicate. restriction on partition columns and cluster columns " +
+    " can be loosed up. f=5 and i=4 are pushed down" in {
+    val result = cc.sql("select * from test_pd where f=5 and i=4").collect()
+  }
+
+  it should "not allow to push down cluster column IN predicate if there is EQ indexed column predicate. " +
+    " i=4 are pushed down" in {
+    val result = cc.sql("select * from test_pd where c=3 and g in (1,2,3) and i=4").collect()
+  }
+
+  it should "not allow to push down partition column IN predicate if there is EQ indexed column predicate. " +
+    " i=4 are pushed down" in {
+    val result = cc.sql("select * from test_pd where c=3 and d in (1,2,3) and i=4").collect()
+  }
+
+  it should "not allow to push down indexed column predicates if all indexed column predicates are none-EQ predicates. " in {
+    val result = cc.sql("select * from test_pd where c=3 and i>4").collect()
+  }
+
+  it should "allow predicate pushdown if there is EQ indexed column predicates. " +
+    " i=4 and j>5 are pushed down" in {
+    val result = cc.sql("select * from test_pd where c=3 and i=4 and j>5").collect()
+  }
+
+  it should "not allow to push down none-primary column IN predicate. " +
+    " e=3 is pushed down" in {
+    val result = cc.sql("select * from test_pd where e=3 and j in (1,2,3)").collect()
+  }
+
+  it should "not allow to push down none-primary column predicates if there are no indexed column predicates. " +
+    " a=1 and b=1 and c=1 and d=3 are pushed down" in {
+    val result = cc.sql("select * from test_pd where a=1 and b=1 and c=1 and d=3 and k=4").collect()
+  }
+}
```

Comment:
Actually we measure the number of rows which were fetched from Cassandra, not the number of rows processed by Spark - so there shall be the less number of rows fetched if the predicate is pushed down, am I right?